pkg/cmd/subcommand/logs.go

```diff
@@ -111,31 +114,114 @@ func (l *Logs) preRun(cf *genericclioptions.ConfigFlags, args []string) error {
 		return err
 	}
-	if f.Status.Serving == nil {
-		return nil
+
+	// Serving stage
+serving:
+	if f.Status.Serving != nil {
+		serving := f.Status.Serving.ResourceRef
+
+		err := l.logsForPods(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("openfunction.io/serving=%s", serving)})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Build stage
+	if f.Status.Build != nil {
+		if buidStageDone {
+			return nil
+		}
+		buidStageDone = true
+		builderRef := f.Status.Build.ResourceRef
+
+		// get openfunction builder ref
+		builder, err := l.functionClient.CoreV1beta1().Builders(f.Namespace).Get(ctx, builderRef, metav1.GetOptions{})
+		if err != nil {
+			statusError, ok := err.(*k8serrors.StatusError)
+			// builder has been cleaned up
+			if ok && statusError.Status().Code == 404 {
+				goto serving
+			} else {
+				return err
+			}
+		}
+
+		// get shipwright builder-buildrun
+		swbuildrunRef := builder.Status.ResourceRef["shipwright.io/buildRun"]
+		swBuildRun, err := l.swClient.ShipwrightV1alpha1().BuildRuns(f.Namespace).Get(ctx, swbuildrunRef, metav1.GetOptions{})
+		if err != nil {
+			statusError, ok := err.(*k8serrors.StatusError)
+			// buildrun has been cleaned up
+			if ok && statusError.Status().Code == 404 {
+				goto serving
+			} else {
+				return err
+			}
+		}
+
+		err = l.logsForPods(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("buildrun.shipwright.io/name=%s", swBuildRun.Name)})
+		if err != nil {
+			return err
+		}
+
+		// reload shipwright buildrun here because we want to know if build succeeded
+		f, err = l.functionClient.CoreV1beta1().Functions(l.namespace).Get(ctx, l.functionName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// only streaming serving logs when build succeeded
+		if l.Follow && f.Status.Build.State == "Succeeded" {
+			goto serving
+		}
 	}
-	serving := f.Status.Serving.ResourceRef
-	listOpt := metav1.ListOptions{LabelSelector: fmt.Sprintf("openfunction.io/serving=%s", serving)}
+
+	return nil
+}
+
+func (l *Logs) logsForPods(ctx context.Context, listOpt metav1.ListOptions) error {
+	// retrieves pod list according to the options
 	podInterface := l.clientSet.CoreV1().Pods(l.namespace)
+	readerList := make([]io.Reader, 0, 5)
 	podList, err := podInterface.List(ctx, listOpt)
 	if err != nil {
 		return err
 	}
-	readerList := make([]io.Reader, 0, len(podList.Items))
+
+	// append stream of each container
 	for _, pod := range podList.Items {
-		logReader, err := podInterface.GetLogs(pod.Name, &corev1.PodLogOptions{Follow: l.Follow, Container: l.containerName}).Stream(ctx)
-		defer logReader.Close()
+		var logReader io.ReadCloser
+		if l.containerName != "" {
+			logReader, err = podInterface.GetLogs(pod.Name, &corev1.PodLogOptions{Follow: l.Follow, Container: l.containerName}).Stream(ctx)
+			if err != nil {
+				return err
+			}
+			defer logReader.Close()
+			readerList = append(readerList, logReader)
+		} else {
+			for _, container := range pod.Spec.Containers {
```

Comment:
There're several kinds of containers in a Function pod:
- the function container whose name is just `function`
- the dapr sidecar container
- the knative sidecar container

Normally we only output logs from the `function` container is enough. This is why we set container name to `function` in https://github.com/OpenFunction/cli/pull/48/commits/21bea4c4b7017c1d728ddb6943cfc942ac8cd154#diff-5c7fb0ae4b5f3afe942a2e3eed411e5a10ffff3e1dccbbf969947d98b79d60d7L54

If the user needs to view logs from dapr sidecar or knative sidecar, he can specify container name. If container name is not specified, the default one `function` will be used.

So we needn't to add a loop here to iterate all containers